### PR TITLE
init.c: convert MyVars to use TAILQ

### LINKED ISSUE #1154: Convert Myvars to use STAILQ
This is an "easy" task ([find more](https://github.com/neomutt/neomutt/issues?q=label:"good%20first%20issue")) and it's open to everyone.  [_read more..._](https://www.neomutt.org/dev/easy-tasks)

## Description

Many NeoMutt structures implement their own linked-list type.  
The Myvars are simply stored as a long linked list.

`STAILQ` is a singly-linked list with pointers to both the head and the tail.  
It's defined in `mutt/queue.h`

### Part 2: Change all users to match

There are plenty of examples in NeoMutt's code of how to iterate through a list.  
**Note**: When deleting items, use `STAILQ_FOREACH_SAFE()`

### PR DESCRIPTION
**Pull request by z2v**
Fixes #1154

### CODE REVIEW

File: init.c [hunk 1]
```diff
@@ -99,44 +99,49 @@ struct MyVar
...
+  struct MyVar *myv = NULL;
   struct MyVar *tmp = NULL;
+  mutt_debug(2, " * myvar_del(%s)\n", NONULL((char *) var));
 
-  for (cur = &MyVars; *cur; cur = &((*cur)->next))
-    if (mutt_str_strcmp((*cur)->name, var) == 0)
-      break;
-
-  if (*cur)
+  STAILQ_FOREACH_SAFE(myv, &MyVars, entries, tmp)
```
Comment (gahr) [inline init.c]: You don’t need the _SAFE version of the macro because you’re not iterating after deletion.

File: init.c [hunk 2]
```diff
@@ -99,44 +99,49 @@ struct MyVar
...
-  struct MyVar **cur = NULL;
-
-  for (cur = &MyVars; *cur; cur = &((*cur)->next))
-    if (mutt_str_strcmp((*cur)->name, var) == 0)
-      break;
+  struct MyVar *myv = NULL;
 
-  if (!*cur)
-    *cur = mutt_mem_calloc(1, sizeof(struct MyVar));
+  mutt_debug(2, " * myvar_set(%s, '<hidden>')\n", NONULL((char *) var));
```
Comment (gahr) [inline init.c]: I’m not sure we want this debug message

File: init.c [hunk 3]
```diff
@@ -4308,11 +4315,16 @@ static int parse_unsubscribe_from(struct Buffer *b, struct Buffer *s,
...
+  struct MyVar *myv = NULL;
 
-  for (cur = MyVars; cur; cur = cur->next)
-    if (mutt_str_strcmp(cur->name, var) == 0)
-      return NONULL(cur->value);
+  STAILQ_FOREACH(myv, &MyVars, entries)
+  {
+    if (mutt_str_strcmp(myv->name, var) == 0)
+    {
+      mutt_debug(2, " * myvar_get(%s)\n", NONULL((char *) var));
```
Comment (gahr) [inline init.c]: Same as above: do we want this debug message?

**Comment by z2v** [comment]
@gahr @flatcap what is the preferred way to make changes in PR? Should I add new commits with changes or squash them in initial commit?

**Comment by flatcap** [comment]
Adding more commits is probably simpler, for now.
That way GitHub will let us see the individual commits, or the combined view.

Ultimately though, it's your branch, so however you want to work is fine.

**Comment by z2v** [comment]
Ok, while reviewing I'll be adding commits. Before merge I'll squash them in single commit. Anyway, replacing STAILQ with TAILQ should be reflected in commit message.

File: init.c [hunk 4]
```diff
@@ -99,44 +99,46 @@ struct MyVar
...
-  struct MyVar **cur = NULL;
-
-  for (cur = &MyVars; *cur; cur = &((*cur)->next))
-    if (mutt_str_strcmp((*cur)->name, var) == 0)
-      break;
-
-  if (!*cur)
-    *cur = mutt_mem_calloc(1, sizeof(struct MyVar));
+  struct MyVar *myv = NULL;
+  struct MyVar *new = NULL;
```
Comment (gahr) [inline init.c]: How about doing without the second variable and just reusing `myv` later one?

Comment (z2v) [reply init.c]: This was the previous version of this code. I've introduced another variable just for clarity. No other reasons.

Comment (z2v) [reply init.c]: Removed.

**Comment by z2v** [comment]
I've squashed commits and fixed commit message.

File: init.c [hunk 5]
```diff
@@ -99,44 +99,45 @@ struct MyVar
 {
   char *name;
   char *value;
-  struct MyVar *next;
+  TAILQ_ENTRY(MyVar) entries;
 };
 
-static struct MyVar *MyVars;
+TAILQ_HEAD(MyVarList, MyVar);
+static struct MyVarList MyVars = TAILQ_HEAD_INITIALIZER(MyVars);
```
Comment (gahr) [inline init.c]: Oh, btw, it's also possible to have the head as an anonymous struct..
```c
static TAILQ_HEAD(, MyVar) MyVars = TAILQ_HEAD_INITIALIZER(MyVars);
```

Comment (z2v) [reply init.c]: Done.

Comment (gahr) [reply init.c]: Thanks!

**Pull request merged**